aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+
+#### get
+
+```proto
+rpc GetPolicyExperiment(GetPolicyExperimentRequest) returns (PolicyExperiment);
+```
+- Standard methods: Get ([AIP-131][aip-131]) **must** be used to get the experiment used for preview mode.
+
+[aip-131]: https://aip.dev/131
+
+#### list
+
+```proto
+rpc ListPolicyExperiments(ListPolicyExperimentsRequest) returns (ListPolicyExperimentsResponse);
+```
+- Standard methods: List ([AIP-132][aip-132]) **should** be used to list experiments for preview mode.
+- List **should** support Filtering ([AIP-160][aip-160]) semantics.
+- Filtering on PreviewMetadata.state **should** indicate which experiments are actively being previewed or not.
+
+[aip-132]: https://aip.dev/132
+[aip-160]: https://aip.dev/160
+
+#### delete
+
+```proto
+rpc DeletePolicyExperiment(DeletePolicyExperimentRequest) returns (google.protobuf.Empty);
+```
+- Standard methods: Delete ([AIP-135][aip-135]) **must** be used to delete experiments.
+
+[aip-135]: https://aip.dev/135
+
+### Parent Resource API Changes
+
+#### delete
+
+- A delete of the top-level resource **must** delete all experiments.
+- To maintain the experiments, the top-level Policy **must** be changed to a no-op Policy instead of using delete.
+
+### Custom Method APIs
+
+#### startPreview
+
+```proto
+rpc StartPreviewPolicyExperiment(StartPreviewPolicyExperimentRequest) returns (StartPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:startPreview"
+    body: "*"
+  }
+}
+
+message StartPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StartPreviewPolicyExperimentResponse {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3;
+  bool successful = 3;
+  string error_message = 4;
+}
+```
+- Admin flow API to start previewing a resource **should** be implemented.
+  - Remains declarative friendly as only an output only field is modified.
+- If startPreview is called specifically on a resource which is already being previewed, then it **should** return an error message.
+- When startPreview is called, the system **should** set values for the log_name, and start_time in PreviewMetadata and set the state to ACTIVE.
+- If startPreview is called on an experiment with an empty set of rules, the deletion of that resource **must** be previewed.
+- startPreview is not idempotent and using it twice on the same resource **should** return an error.
+- If the API has an error, no fields in PreviewMetadata **should** change.
+- Etags **must** be used to guarantee resource freshness.
+
+#### stopPreview
+
+```proto
+rpc StopPreviewPolicyExperiment(StopPreviewPolicyExperimentRequest) returns (StopPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:stopPreview"
+    body: "*"
+  }
+}
+
+message StopPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StopPreviewPolicyExperimentResponse {
```

Comment:
Same note as `StartPreviewPolicyExperimentResponse` message --- consider returning the PolicyExperiment as, or an LRO-of-PolicyExperiment if the stop/cancel method needs to be long-running.

Reply:
done